### ISSUE TITLE
Multiple producers for Ninja target

### ISSUE THREAD

**Issue by zajo** [mesonbuild/meson#1813]
Run 'meson build' in the p1 project. This leads to an object file name clash for the object names deduced from f1.cpp and f1.cpp, one of which is in the project p1, the one in its subproject p2.
[p1.zip](https://github.com/mesonbuild/meson/files/1008733/p1.zip)

**Comment by nirbheek** [comment]
Hmm, we should actually handle properly this for intermediate objects. It can be very common to have, say, a file called `main.c` in two totally different subprojects or even subdirs.

**Comment by zajo** [comment]
Yes. One way is to detect the clash and add a numerical suffix to subsequently generated object names. Alternatively, the issue will go away if subproject object file names are prepended with the subproject's name. I prefer the former. I was going to try to fix it but I'm not a Python programmer and decided to give up when I found myself googling 'string operations in python'. :)

**Comment by nirbheek** [comment]
Ah, this is because you added `f1.cpp` to `sources:` in `declare_dependency()`. This will cause it to be built in the current directory. We append the full path to the intermediate object, so this is usually not a problem.

**Comment by nirbheek** [comment]
This is what the (invalid) `build.ninja` would contain if Meson were to generate it:

```ninja
build bug@exe/f1.cpp.o: cpp_COMPILER ../f1.cpp
 DEPFILE = bug@exe/f1.cpp.o.d
 ARGS = '-Ibug@exe' '-I.' '-I..' '-pipe'
build bug@exe/f1.cpp.o: cpp_COMPILER ../subprojects/p2/f1.cpp
 DEPFILE = bug@exe/f1.cpp.o.d
 ARGS = '-Ibug@exe' '-I.' '-I..' '-pipe'
```

**Comment by zajo** [comment]
Looking at this perhaps the cleaner fix is to prepend subproject names: ../subprojects/p2/f1.cpp -> p2_f1.cpp.o (while still using just f1.cpp.o for the one originating in the main project.)

**Comment by nirbheek** [comment]
We should probably do `subprojects_p2_f1.cpp` just to minimize the chances of a collision even further; just in case there's a subdir `p2` with a file `f1.cpp` in the current project!
